File: e2e/fixtures/research.ts
/**
 * Canned research responses.
 *
 * The panel's states — needs_browser, robots_denied, budget_exceeded — depend
 * on what a real website does, which an offline suite cannot arrange. These
 * fixtures are served through `page.route`, so the tests drive the frontend
 * deterministically without the backend ever fetching anything.
 *
 * The shapes mirror the API contract exactly; a drift between them shows up as
 * a TypeScript-free runtime mismatch, which the "applied payload" assertions
 * would catch.
 */

export const RESEARCH_RUN_ID = "11111111-2222-3333-4444-555555555555";

export interface ResearchRunFixture {
  research_id: string;
  company_id: string | null;
  company_name: string;
  website: string;
  status: "completed" | "partial" | "failed";
  failure_code: string | null;
  started_at: string;
  completed_at: string | null;
  pages_fetched: number;
  pages_failed: number;
  claims_extracted: number;
  claims_validated: number;
  extractor: { provider: string; model: string; prompt_version: string } | null;
  profile: Record<string, unknown>;
  pages: Array<Record<string, unknown>>;
  claims: Array<Record<string, unknown>>;
  rejected_claims: Array<Record<string, unknown>>;
  warnings: string[];
  unknown_dimensions: string[];
  output_language: string;
  action?: string;
}

const NOW = "2026-07-19T12:00:00Z";

function page(position: number, path: string, reason: string) {
  return {
    position,
    url: `https://acme.example${path}`,
    final_url: `https://acme.example${path}`,
    http_status: 200,
    content_type: "text/html",
    fetched_at: NOW,
    content_chars: 812,
    truncated: false,
    discovery_reason: reason,
  };
}

function claim(
  position: number,
  kind: string,
  detail: string,
  path: string,
  confidence: number,
) {
  return {
    position,
    kind,
    detail,
    evidence_snippet: `verbatim sentence supporting ${kind}`,
    source_url: `https://acme.example${path}`,
    confidence,
  };
}

const BASE: ResearchRunFixture = {
  research_id: RESEARCH_RUN_ID,
  company_id: null,
  company_name: "Acme Hardware",
  website: "https://acme.example",
  status: "completed",
  failure_code: null,
  started_at: NOW,
  completed_at: NOW,
  pages_fetched: 2,
  pages_failed: 0,
  claims_extracted: 3,
  claims_validated: 3,
  extractor: {
    provider: "fake",
    model: "fake-research-v1",
    prompt_version: "research-extract-fake-v1",
  },
  profile: {
    summary: "Hardware importer",
    industry: null,
    products: [],
    locations: [],
    size_hint: null,
    year_founded: null,
    mentions_importing: true,
  },
  pages: [page(0, "/", "homepage"), page(1, "/about", "ranked:about")],
  claims: [
    claim(0, "import_activity", "进口五金，来自亚洲", "/", 0.8),
    claim(1, "company_scale", "自有仓库约 12 万平方英尺", "/", 0.6),
    claim(2, "growth_signal", "进口量逐年增长", "/about", 0.7),
  ],
  rejected_claims: [
    {
      reason: "snippet_not_found",
      kind: "cargo_value_potential",
      detail: "invented value claim",
      warning: "claim rejected (snippet_not_found): not present on page",
    },
  ],
  warnings: [],
  unknown_dimensions: ["shipping_fit", "pain_point"],
  output_language: "zh-CN",
  action: "recorded",
};

export const COMPLETED_RUN: ResearchRunFixture = { ...BASE };

export const PARTIAL_RUN: ResearchRunFixture = {
  ...BASE,
  status: "partial",
  failure_code: null,
  pages_fetched: 1,
  pages_failed: 1,
  pages: [page(0, "/", "homepage")],
  warnings: ["https://acme.example/about could not be read (http_error): HTTP 404"],
  action: "partial",
};

export const NEEDS_BROWSER_RUN: ResearchRunFixture = {
  ...BASE,
  status: "partial",
  failure_code: "needs_browser",
  claims: [],
  claims_extracted: 0,
  claims_validated: 0,
  rejected_claims: [],
  warnings: ["pages yielded almost no text — the site likely requires a browser"],
  action: "partial",
};

export const ROBOTS_DENIED_RUN: ResearchRunFixture = {
  ...BASE,
  status: "failed",
  failure_code: "robots_denied",
  pages: [],
  pages_fetched: 0,
  claims: [],
  claims_extracted: 0,
  claims_validated: 0,
  rejected_claims: [],
  warnings: ["robots.txt disallows the homepage: https://acme.example"],
  action: "failed",
};

export const BUDGET_EXCEEDED_RUN: ResearchRunFixture = {
  ...BASE,
  status: "partial",
  failure_code: "budget_exceeded",
  warnings: ["time budget exhausted — extracting from pages read so far"],
  action: "partial",
};

/** What confirm returns for a run with no company: the form payload. */
export function confirmResponse(
  signals: Array<{ kind: string; detail: string }>,
  sources: Array<{ source: string; reference: string }>,
) {
  return {
    research_id: RESEARCH_RUN_ID,
    action: "recorded",
    company_id: null,
    summary: {
      accepted: signals.length,
      edited: 0,
      rejected: 0,
      total: signals.length,
    },
    promotions: signals.map((signal, index) => ({
      claim_position: index,
      decision: "accepted",
      kind: signal.kind,
      detail: signal.detail,
      company_source_position: null,
      company_signal_position: null,
      source_reused: false,
      idempotent: false,
    })),
    application_payload: {
      company_name: "Acme Hardware",
      website: "https://acme.example",
      sources,
      signals,
    },
    warnings: [],
  };
}

File: apps/backend/app/database/mappers/research.py
"""ResearchRun aggregate ↔ persistence mapping."""

from dataclasses import asdict
from typing import Any
from uuid import UUID

from app.database.models.research import (
    ResearchClaimModel,
    ResearchPageModel,
    ResearchPromotionModel,
    ResearchRunModel,
)
from app.domain.research import (
    ClaimRejectionReason,
    ExtractorIdentity,
    PromotionDecision,
    RejectedClaim,
    ResearchClaim,
    ResearchFailureCode,
    ResearchPage,
    ResearchProfile,
    ResearchPromotion,
    ResearchRun,
    ResearchRunStatus,
)


class ResearchRunMapper:
    @staticmethod
    def to_model(run: ResearchRun) -> ResearchRunModel:
        extractor = run.extractor
        return ResearchRunModel(
            id=run.id,
            company_id=run.company_id,
            company_name=run.company_name,
            website=run.website,
            status=run.status.value,
            failure_code=run.failure_code.value if run.failure_code else None,
            started_at=run.started_at,
            completed_at=run.completed_at,
            pages_fetched=run.pages_fetched,
            pages_failed=run.pages_failed,
            claims_extracted=run.claims_extracted,
            claims_validated=run.claims_validated,
            extractor_provider=extractor.provider if extractor else None,
            extractor_model=extractor.model if extractor else None,
            prompt_version=extractor.prompt_version if extractor else None,
            profile_json=asdict(run.profile),
            warnings_json=list(run.warnings),
            rejected_json=[
                {
                    "reason": rejection.reason.value,
                    "kind": rejection.kind,
                    "detail": rejection.detail,
                    "warning": rejection.warning,
                }
                for rejection in run.rejected_claims
            ],
            pages=[
                ResearchPageModel(
                    research_id=run.id,
                    position=page.position,
                    url=page.url,
                    final_url=page.final_url,
                    http_status=page.http_status,
                    content_type=page.content_type,
                    fetched_at=page.fetched_at,
                    content_chars=page.content_chars,
                    bytes_read=page.bytes_read,
                    truncated=page.truncated,
                    discovery_reason=page.discovery_reason,
                )
                for page in run.pages
            ],
            claims=[
                ResearchClaimModel(
                    research_id=run.id,
                    position=claim.position,
                    kind=claim.kind,
                    detail=claim.detail,
                    evidence_snippet=claim.evidence_snippet,
                    source_page_position=claim.source_page_position,
                    confidence=claim.confidence,
                )
                for claim in run.claims
            ],
            promotions=[
                ResearchPromotionModel(
                    research_id=run.id,
                    claim_position=promotion.claim_position,
                    decision=promotion.decision.value,
                    reviewed_at=promotion.reviewed_at,
                    reviewer_name=promotion.reviewer_name,
                    edited_detail=promotion.edited_detail,
                    company_id=(
                        promotion.company_id
                        if isinstance(promotion.company_id, UUID)
                        else None
                    ),
                    company_signal_position=promotion.company_signal_position,
                )
                for promotion in run.promotions
            ],
        )

    @staticmethod
    def to_domain(model: ResearchRunModel) -> ResearchRun:
        run = ResearchRun(
            id=model.id,
            company_id=model.company_id,
            company_name=model.company_name,
            website=model.website,
            started_at=model.started_at,
        )
        run._status = ResearchRunStatus(model.status)
        run._failure_code = (
            ResearchFailureCode(model.failure_code) if model.failure_code else None
        )
        run._completed_at = model.completed_at
        run._pages_failed = model.pages_failed
        run._claims_extracted = model.claims_extracted
        run._warnings = list(model.warnings_json or [])
        run._rejected = [
            RejectedClaim(
                reason=ClaimRejectionReason(entry["reason"]),
                kind=entry["kind"],
                detail=entry["detail"],
                warning=entry["warning"],
            )
            for entry in (model.rejected_json or [])
        ]
        run._profile = _profile_from_json(model.profile_json or {})
        if model.extractor_provider and model.extractor_model and model.prompt_version:
            run._extractor = ExtractorIdentity(
                provider=model.extractor_provider,
                model=model.extractor_model,
                prompt_version=model.prompt_version,
            )
        run._pages = [
            ResearchPage(
                position=page.position,
                url=page.url,
                final_url=page.final_url,
                http_status=page.http_status,
                content_type=page.content_type,
                fetched_at=page.fetched_at,
                content_chars=page.content_chars,
                bytes_read=page.bytes_read,
                truncated=page.truncated,
                discovery_reason=page.discovery_reason,
            )
            for page in model.pages
        ]
        run._claims = [
            ResearchClaim(
                position=claim.position,
                kind=claim.kind,
                detail=claim.detail,
                evidence_snippet=claim.evidence_snippet,
                source_page_position=claim.source_page_position,
                confidence=claim.confidence,
            )
            for claim in model.claims
        ]
        run._promotions = [
            ResearchPromotion(
                claim_position=promotion.claim_position,
                decision=PromotionDecision(promotion.decision),
                reviewed_at=promotion.reviewed_at,
                reviewer_name=promotion.reviewer_name,
                edited_detail=promotion.edited_detail,
                company_id=promotion.company_id,
                company_signal_position=promotion.company_signal_position,
            )
            for promotion in model.promotions
        ]
        return run


def _profile_from_json(payload: dict[str, Any]) -> ResearchProfile:
    return ResearchProfile(
        summary=payload.get("summary"),
        industry=payload.get("industry"),
        products=tuple(payload.get("products") or ()),
        locations=tuple(payload.get("locations") or ()),
        size_hint=payload.get("size_hint"),
        year_founded=payload.get("year_founded"),
        mentions_importing=payload.get("mentions_importing"),
    )
